farmerbot/internal/power_manager.go

```diff
@@ -0,0 +1,399 @@
+		}
+	}
+
+	var possibleNodes []node
+	for _, node := range p.state.nodes {
+		gpus := node.gpus
+		if nodeOptions.HasGPUs > 0 {
+			if len(nodeOptions.GPUVendors) > 0 {
+				gpus = filterGPUs(gpus, nodeOptions.GPUVendors, false)
+			}
+
+			if len(nodeOptions.GPUDevices) > 0 {
+				gpus = filterGPUs(gpus, nodeOptions.GPUDevices, true)
+			}
+
+			if len(gpus) < int(nodeOptions.HasGPUs) {
+				continue
+			}
+		}
+
+		if nodeOptions.Certified && !node.Certification.IsCertified {
+			continue
+		}
+
+		if nodeOptions.PublicConfig && !node.PublicConfig.HasValue {
+			continue
+		}
+
+		if node.hasActiveRentContract {
+			continue
+		}
+
+		if nodeOptions.Dedicated {
+			if !node.dedicated || !node.isUnused() {
+				continue
+			}
+		} else {
+			if node.dedicated && nodeOptionsCapacity != node.resources.total {
+				continue
+			}
+		}
+
+		if slices.Contains(nodeOptions.NodeExclude, uint32(node.ID)) {
+			continue
+		}
+
+		if !node.canClaimResources(nodeOptionsCapacity, p.state.config.Power.OverProvisionCPU) {
+			continue
+		}
+
+		possibleNodes = append(possibleNodes, node)
+	}
+
+	if len(possibleNodes) == 0 {
+		return 0, fmt.Errorf("could not find a suitable node with the given options: %+v", possibleNodes)
+	}
+
+	// Sort the nodes on power state (the ones that are ON first then waking up, off, shutting down)
+	sort.Slice(possibleNodes, func(i, j int) bool {
+		return possibleNodes[i].powerState < possibleNodes[j].powerState
+	})
+
+	nodeFound := possibleNodes[0]
+	log.Debug().Str("manager", "node").Uint32("node ID", uint32(nodeFound.ID)).Msg("Found a node")
+
+	// claim the resources until next update of the data
+	// add a timeout (after 30 minutes we update the resources)
+	nodeFound.timeoutClaimedResources = time.Now().Add(timeoutPowerStateChange)
+
+	if nodeOptions.Dedicated || nodeOptions.HasGPUs > 0 {
+		// claim all capacity
+		nodeFound.claimResources(nodeFound.resources.total)
+	} else {
+		nodeFound.claimResources(nodeOptionsCapacity)
+	}
+
+	// claim public ips until next update of the data
+	if nodeOptions.PublicIPs > 0 {
+		nodeFound.publicIPsUsed += nodeOptions.PublicIPs
+	}
+
+	// power on the node if it is down or if it is shutting down
+	if nodeFound.powerState == off || nodeFound.powerState == shuttingDown {
+		if err := p.PowerOn(sub, uint32(nodeFound.ID)); err != nil {
+			return 0, fmt.Errorf("failed to power on found node %d", nodeFound.ID)
+		}
+	}
+
+	// update claimed resources
+	err := p.state.updateNode(nodeFound)
+	if err != nil {
+		return 0, fmt.Errorf("failed to power on found node %d", nodeFound.ID)
+	}
+
+	return uint32(nodeFound.ID), nil
+}
+
+// PeriodicWakeUp for waking up nodes daily
+func (p *PowerManager) PeriodicWakeUp(sub Sub) error {
+	now := time.Now()
+
+	offNodes := p.state.filterNodesPower([]powerState{off})
+
+	periodicWakeUpStart := p.state.config.Power.PeriodicWakeUpStart.PeriodicWakeUpTime()
+
+	var wakeUpCalls uint8
+	nodesLen := len(p.state.nodes)
+
+	for _, node := range offNodes {
```

Comment:
as i already mentioned in the previous review please this can done on `node by node` bases in the main loop. for example why this instead doesn't become 
```
 /// THIS IS MAIN LOOP NOT THIS LOOP, THE ONE IN farmertbo.Run
  for node in nodes {
         node = getLatestState() 
          f.periodicWakeUp(&node) // this one decides if node to be powerd on or not
   } 
``